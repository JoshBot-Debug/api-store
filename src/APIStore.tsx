import React, { createContext, PropsWithChildren, useEffect, useMemo, useReducer } from "react";
import reducer, { prepareInitialState } from "./reducer";
import { Model, UseAPIStore } from "./types";

type Context = UseAPIStore.Context<{}>;

const defaultContext: Context = {
  cache: {},
  upsert: () => { },
  get: () => null,
}

export const APIStoreContext = createContext<Context>(defaultContext)

interface APIStoreProps extends PropsWithChildren {
  model: Model.Model;
  debug?: boolean;
}

export default function APIStore(props: APIStoreProps) {

  const { children, debug } = props;

  const model = props.model as Model.Proto<Model.Table.Created>;

  const initialState = prepareInitialState(model);

  const [state, dispatch] = useReducer(reducer, initialState);

  useEffect(() => {
    if (!debug) return;
    console.log(state.cache)
  }, [state])

  const upsert: Context["upsert"] = (payload) => dispatch({ type: "upsert", payload, model });

  const get: Context["get"] = (table, where, fields) => {
    if (!where) return null;
    // TODO need to find out where the value is mutating and remove this.
    return model.get(table, state.cache, where, fields);
  }

  const value = useMemo<Context>(() => ({
    cache: state.cache,
    upsert,
    get,
  }), [state.cache])

  return <APIStoreContext.Provider value={value}>{children}</APIStoreContext.Provider>
}