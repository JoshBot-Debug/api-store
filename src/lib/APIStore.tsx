import React, { createContext, PropsWithChildren, useMemo, useReducer } from "react";
import reducer, { prepareInitialState } from "./reducer";

type Context = UseAPIStore.Context<{}>;

const defaultContext: Context = {
  cache: {},
  upsert: () => { },
  get: () => null,
}

export const APIStoreContext = createContext<Context>(defaultContext)

interface APIStoreProps extends PropsWithChildren {
  model: Model.Model
}

export default function APIStore(props: APIStoreProps) {

  const { children } = props;

  const model = props.model as Model.Proto<Model.Table.Created>;

  const initialState = prepareInitialState(model);

  const [state, dispatch] = useReducer(reducer, initialState);

  const upsert: Context["upsert"] = (payload) => dispatch({ type: "upsert", payload, model });

  const get: Context["get"] = (table, where, fields) => {
    if (!where) return null;
    return model.get(table, state.cache, where, fields);
  }

  const value = useMemo<Context>(() => ({
    cache: state.cache,
    upsert,
    get,
  }), [state.cache])

  return <APIStoreContext.Provider value={value}>{children}</APIStoreContext.Provider>
}