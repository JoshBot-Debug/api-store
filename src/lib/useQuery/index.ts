import { useContext, useEffect, useMemo, useReducer, useState } from "react";
import { APIStoreContext } from "../Provider";
import { initialState, reducer } from "./reducer";

export function useQuery<Data>(config: UseAPIStore.Config<Data>): UseAPIStore.UseQueryReturn<Data> {

  const {
    table,
    enabled = true,
    fields = null,
  } = config;

  const {
    result: _result,
    fetch,
    where: _where
  } = config.get;

  const [state, dispatch] = useReducer(reducer, initialState);

  const context = useContext<UseAPIStore.Context<Data>>(APIStoreContext as any);

  const [where, setWhere] = useState<UseAPIStore.WhereClause<Data> | null>(_where ?? _result ?? null);

  useEffect(() => { onMount(); }, []);

  async function onMount() {
    try {
      if (_result) {
        context.upsert({ table, data: _result });
        setWhere(_where ?? _result);
      }
      if (enabled && fetch) {
        dispatch({ type: "isLoading", payload: true });
        const data = await fetch();
        setWhere(_where ?? data)
        context.upsert({ table, data });
      }
    }
    catch (error) {
      dispatch({ type: "error", payload: (error as any).message ?? `Something went wrong... \nERROR: ${JSON.stringify(error)}` })
      throw error;
    }
    finally {
      dispatch({ type: "isLoading", payload: false });
    }
  }

  async function refetch(...args: any[]) {
    if (!fetch) throw new Error("You cannot call refetch without passing a fetch function to the hook.");
    try {
      dispatch({ type: "isFetching", payload: true });
      const data = await fetch(...args);
      setWhere(_where ?? data)
      context.upsert({ table, data });
      return data;
    }
    catch (error) {
      dispatch({ type: "error", payload: (error as any).message ?? `Something went wrong... \nERROR: ${JSON.stringify(error)}` })
      throw error;
    }
    finally { dispatch({ type: "isFetching", payload: false }); }
  }

  const result = JSON.stringify(context.get(table, where, fields));

  return useMemo(() => ({
    result: JSON.parse(result),
    refetch,
    ...state,
  }), [state.isFetching, state.error, result])
}