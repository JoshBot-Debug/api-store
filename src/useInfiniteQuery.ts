import { useEffect, useMemo, useState } from "react";
import { UseAPIStore } from "./types";
import { useStore, useStoreIndex } from "./store";
import { withOptions } from "@jjmyers/object-relationship-store";

export function useInfiniteQuery<
  I extends string,
  N extends string,
  O extends Record<string, any>,
  R extends Record<string, any>,
  NextPageParams,
>(config: UseAPIStore.UseInfiniteQueryConfig<I, N, O, R, NextPageParams>, deps: any[] = []): UseAPIStore.UseInfiniteQueryReturn<O, R, NextPageParams> {

  const {
    index,
    select,
    enabled = true,
    getNextPageParams,
    fetch,
    getData
  } = config;

  const store = useStore();
  const state = useStoreIndex(index, select);
  const [error, setError] = useState<any | null>(null);
  const [isLoading, setIsLoading] = useState(false);
  const [isFetching, setIsFetching] = useState(false);
  const [nextPageParams, setNextPageParams] = useState<NextPageParams | null>(null);
  const [hasNextPage, setHasNextPage] = useState(true);

  useEffect(() => {
    if (!enabled || !fetch) return;
    refresh(false);
  }, [config.enabled, ...deps]);

  
  async function refresh(clearIndex: boolean = true) {
    if (!enabled || !fetch) return null;
    setIsLoading(true);
    setNextPageParams(null);
    setHasNextPage(true);
    setIsFetching(false);
    setError(null);
    const result = await makeRequest([], !!clearIndex);
    setIsLoading(false);
    return result;
  }

  async function makeRequest(args: any[], clearIndex?: boolean) {
    if (!hasNextPage) return null;
    if (!fetch) throw new Error("You cannot call refetch without passing a fetch function to the hook.");
    try {
      setIsFetching(true);
      const arg = args.length > 0 ? args : [nextPageParams];
      const result = await fetch(...arg);
      const nextParams = getNextPageParams(result);
      const data = getData ? getData(result) : result as unknown as O[];
      if (nextParams) { setNextPageParams(nextParams); }
      else {
        setNextPageParams(null);
        setHasNextPage(false);
      }
      if (clearIndex) store.destroy(index);
      store.upsert(withOptions(data, { __indexes__: [index] }));
      return { result, data };
    }
    catch (error) {
      setError(error);
      setNextPageParams(null);
      setHasNextPage(false);
    }
    finally { setIsFetching(false); }
    return null;
  }


  async function fetchNextPage(...args: any[]) {
    return await makeRequest(args);
  }


  return useMemo(() => ({
    state,
    refresh,
    fetchNextPage,
    nextPageParams,
    isLoading,
    hasNextPage,
    isFetching,
    error,
  }), [isLoading, hasNextPage, nextPageParams, isFetching, error, state])
}