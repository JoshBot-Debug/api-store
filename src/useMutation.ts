import { useMemo, useState } from "react";
import { UseAPIStore } from "./types";
import { useStore } from "./store";

export function useMutation<
  R extends Record<string, any>,
  A extends Array<any>
>(
  config: UseAPIStore.UseMutationConfig<R, A>,
  deps: any[] = [],
): UseAPIStore.UseMutationReturn<R, A> {

  const {
    mutate
  } = config;

  const store = useStore();
  const [error, setError] = useState<any | null>(null);
  const [isLoading, setIsLoading] = useState(false);

  async function makeMutation(...args: A) {
    try {
      setIsLoading(true);
      let result = null;
      try { result = await mutate(...args); }
      catch (error) { setError(String(error)); }
      store.upsert(result);
      return result;
    }
    finally { setIsLoading(false); }

    return null;
  }

  return useMemo(() => ({
    error,
    isLoading,
    mutate: makeMutation,
  }), [isLoading, error, ...deps])
}