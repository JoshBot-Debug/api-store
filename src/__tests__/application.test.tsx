import React from "react";
import { render, renderHook, screen, waitFor } from "@testing-library/react";
import APIStore from "../APIStore";
import { fakeFetch, fakePaginatingFetch, useRenderCount } from "./test-utils";
import { createTable, createModel } from "../model";
import { useQuery } from "../useQuery";
import { act } from "react-dom/test-utils";
import { useMutation } from "../useMutation";
import { useInfiniteQuery } from "../useInfiniteQuery";
import { UseAPIStore } from "../types";

/**
 * The purpose of this file is to test a real world situation.
 * 
 * We will have :~
 * 1. An authenticated user.
 * 2. Other users.
 * 3. Posts
 * 4. Comments
 * 5. Profiles
 * 
 * 
 * Scene 1
 * Sign up page
 * 
 * A user enters:
 *  username, email, password, gender, dob
 * 
 * The response is:
 *  token, user information.
 * 
 * The result:
 *  The user is authenticated.
 * 
 */


type User = {
  id: number;
  username: string;
  email: string;
  gender: string;
  dob: string;
  token?: Token | number;
}

type Token = {
  user: User | number;
  token: string;
}

const token = createTable("token", {
  token: "string",
}, { pk: "user" })

const user = createTable("user", {
  id: "number",
  username: "string",
  email: "string",
  gender: "string",
  dob: "string",
});

const image = createTable("image", {
  id: "number",
  aspectRatio: "number",
})

const thumbnail = createTable("thumbnail", {
  id: "number",
  src: "string",
})

token.hasOne(user)

user.hasOne(token)
user.hasOne(image, "profileImage")
user.hasOne(image, "bannerImage")

image.hasMany(thumbnail, "thumbnails")


const model = createModel([
  user,
  token,
  image,
  thumbnail,
])

const users = {
  10: {
    id: 10,
    username: "Joshua",
    email: "joshua@gmail.com",
    gender: "male",
    dob: "21-02-1998",
    token: {
      user: 10,
      token: "<secret>"
    }
  },
  11: {
    id: 11,
    username: "Hannah",
    email: "hannah@gmail.com",
    gender: "female",
    dob: "26-01-2000",
    token: 11,
  },
  12: {
    id: 12,
    username: "May",
    email: "amy@gmail.com",
    gender: "female",
    dob: "26-01-2000",
    token: 12,
  },
}

const wrapper = ({ children }: React.PropsWithChildren) => {


  function GetOtherUsers() {

    useQuery({
      table: "user",
      get: { result: Object.values(users) },
    });

    return (<>{children}</>)
  }

  return <APIStore model={model}><GetOtherUsers /></APIStore>
}

describe("useQuery hook tests", () => {


  it("useQuery with fetcher and result.", async () => {

    const data = {
      id: 10,
      username: "Joshua",
      email: "joshua@gmail.com",
      gender: "male",
      dob: "21-02-1998",
      token: {
        user: 10,
        token: "<secret>",
      }
    }

    const updatedData = { ...data, dob: "<updated-dob>" };

    const fetch = () => fakeFetch(updatedData, 100)

    const { result } = renderHook(() => (
      useQuery({
        table: "user",
        get: {
          result: data,
          fetch,
        },
      })
    ), { wrapper });

    expect(result.current.isFetching).toBe(true);
    expect(result.current.result).toStrictEqual(data);

    await waitFor(() => expect(result.current.isFetching).toBe(false))

    expect(result.current.isFetching).toBe(false);
    expect(result.current.result).toStrictEqual(updatedData);

  });


  it("useQuery with fetcher", async () => {

    const data = {
      id: 10,
      username: "Joshua",
      email: "joshua@gmail.com",
      gender: "male",
      dob: "21-02-1998",
      token: {
        user: 10,
        token: "<secret>",
      }
    }

    const fetch = () => fakeFetch(data, 100)

    const { result } = renderHook(() => (
      useQuery({
        table: "user",
        get: { fetch },
      })
    ), { wrapper });

    expect(result.current.isFetching).toBe(true);

    await waitFor(() => expect(result.current.isFetching).toBe(false))

    expect(result.current.isFetching).toBe(false);
    expect(result.current.result).toStrictEqual(data);

  });


  it("useQuery with result", async () => {

    const data = {
      id: 10,
      username: "Joshua",
      email: "joshua@gmail.com",
      gender: "male",
      dob: "21-02-1998",
      token: {
        user: 10,
        token: "<secret>",
      }
    }

    const { result } = renderHook(() => (
      useQuery({
        table: "user",
        get: { result: data },
      })
    ), { wrapper });

    expect(result.current.result).toStrictEqual(data)

  });


  it("useQuery with result and where clause, token.user", async () => {

    const data = {
      token: "<secret>",
      user: {
        id: 10,
        username: "Joshua",
        email: "joshua@gmail.com",
        gender: "male",
        dob: "21-02-1998",
      },
    }

    const { result } = renderHook(() => (
      useQuery({
        table: "token",
        get: {
          result: data,
          where: {
            user: {
              id: 10
            }
          }
        },
      })
    ), { wrapper });

    expect(result.current.result).toStrictEqual({ ...data, user: { ...data.user, token: 10 } })
  });


  it("useQuery with result and where clause, user.token", async () => {

    const data = {
      id: 10,
      username: "Joshua",
      email: "joshua@gmail.com",
      gender: "male",
      dob: "21-02-1998",
      token: {
        user: 10,
        token: "<secret>",
      }
    }

    const { result } = renderHook(() => (
      useQuery({
        table: "user",
        get: {
          result: data,
          where: {
            id: 10,
            token: {
              user: 10
            }
          }
        },
      })
    ), { wrapper });


    expect(result.current.result).toStrictEqual(data)
  });


  it("useQuery data is token with user, where clause get only token.", async () => {

    const data = {
      token: "<secret>",
      user: {
        id: 10,
        username: "Joshua",
        email: "joshua@gmail.com",
        gender: "male",
        dob: "21-02-1998",
      },
    }

    const { result } = renderHook(() => (
      useQuery({
        table: "token",
        get: {
          result: data,
          where: {
            user: 10
          }
        },
      })
    ), { wrapper });

    expect(result.current.result).toStrictEqual({
      user: 10,
      token: "<secret>",
    });

  });


  it("useQuery data is user with token, where clause get only user.", async () => {

    const data: User = {
      id: 10,
      username: "Joshua",
      email: "joshua@gmail.com",
      gender: "male",
      dob: "21-02-1998",
      token: {
        user: 10,
        token: "<secret>"
      }
    }

    const { result } = renderHook(() => (
      useQuery({
        table: "user",
        get: {
          result: data,
          where: { id: 10 }
        },
      })
    ), { wrapper });


    expect(result.current.result).toStrictEqual({
      id: 10,
      username: "Joshua",
      email: "joshua@gmail.com",
      gender: "male",
      dob: "21-02-1998",
      token: 10
    });

  });


  it("useQuery get a single existing user Hannah by id.", async () => {

    const { result } = renderHook(() => (
      useQuery({
        table: "user",
        get: {
          where: { id: 11 }
        },
      })
    ), { wrapper });

    expect(result.current.result).toStrictEqual(users[11]);

  });


  it("useQuery get an array existing users by gender.", async () => {

    const { result } = renderHook(() => (
      useQuery({
        table: "user",
        get: {
          where: { gender: "female" }
        },
      })
    ), { wrapper });

    expect(result.current.result).toStrictEqual([users[11], users[12]]);
  });

  it("useQuery result is an array.", async () => {

    const { result } = renderHook(() => (
      useQuery({
        table: "user",
        get: { result: Object.values(users) },
      })
    ), { wrapper });

    expect(result.current.result).toStrictEqual(Object.values(users));
  });


  it("useQuery result is an array, where has duplicate values.", async () => {

    const { result } = renderHook(() => (
      useQuery({
        table: "user",
        get: {
          result: Object.values(users),
          where: [...Object.values(users), ...Object.values(users)]
        },
      })
    ), { wrapper });


    expect(result.current.result).toStrictEqual(Object.values(users));
  });


  it("useQuery getData function test.", async () => {

    const { result } = renderHook(() => (
      useQuery({
        table: "user",
        get: {
          result: { data: Object.values(users) },
        },
        getData: r => r.data
      })
    ), { wrapper });

    expect(result.current.result).toStrictEqual(Object.values(users));
  });

});


describe("useQuery hook component tests", () => {


  function Dob(props: { id: number; }) {

    const store = useQuery<User, User>({
      table: "user",
      get: {
        where: { id: props.id }
      },
      fields: { user: ["dob"] },
      enabled: false
    })

    const count = useRenderCount([store])

    return (
      <>
        <p data-testid={`count-dob-${props.id}`} >Count({props.id}): {count}</p>
        <p data-testid="profile-dob">{store.result?.dob}</p>
      </>
    )
  }

  function Profile(props: { id: number; }) {

    const store = useQuery<User, User>({
      table: "user",
      get: {
        where: { id: props.id }
      },
      enabled: false,
      fields: {
        user: ["username"]
      }
    })

    const count = useRenderCount([store])

    return (
      <>
        <p data-testid={`count-profile-${props.id}`} >Count({props.id}): {count}</p>
        <p data-testid="profile-username">{store.result?.username}</p>
        <Dob id={props.id} />
      </>
    )
  }

  const UserComponent = (props: { id: number; }) => {

    const data = (users as any)[props.id];

    const fetch = (rand: any) => fakeFetch({ ...data, dob: `<updated-dob> [${rand}]` }, 100)

    const store = useQuery({
      table: "user",
      get: {
        result: data,
        fetch
      },
      fields: {
        user: ["dob"]
      },
      enabled: false
    })

    const count = useRenderCount([store])

    const handleButtonClick = () => store.refetch(count);

    return (
      <div>
        <p data-testid={`dob-${props.id}`} >{store.result?.dob}</p>
        <p data-testid={`count-${props.id}`} >Count({props.id}): {count}</p>
        <button data-testid={`refetch-${props.id}`} onClick={handleButtonClick}>Update {props.id}</button>
      </div>
    );
  }


  it("Using 'result': Should rerender only the component where data changes", async () => {

    function UserListComponent() {

      const store = useQuery({
        table: "user",
        get: { result: Object.values(users) },
        fields: {
          user: ["id", "dob", "username"],
        }
      })

      const count = useRenderCount([store])

      const renderItem = (user: User, index: number) => user && <UserComponent key={index} id={user.id} />

      return (
        <>
          <p data-testid={`count-list`} >Count(list): {count}</p>
          {store.result?.map(renderItem)}
        </>
      )
    }

    render(
      <>
        <UserListComponent />
        <Profile id={12} />
      </>,
      { wrapper }
    )

    const countListElement = screen.getByTestId("count-list");
    const count10Element = screen.getByTestId("count-10");
    const count11Element = screen.getByTestId("count-11");
    const count12Element = screen.getByTestId("count-12");
    const count12Dob = screen.getByTestId("dob-12");

    const countProfileUsername = screen.getByTestId("count-profile-12");
    const countProfileDob = screen.getByTestId("count-dob-12");

    expect(countListElement.textContent).toBe('Count(list): 2');
    expect(count10Element.textContent).toBe('Count(10): 1');
    expect(count11Element.textContent).toBe('Count(11): 1');
    expect(count12Element.textContent).toBe('Count(12): 1');

    expect(countProfileUsername.textContent).toBe('Count(12): 2');
    expect(countProfileDob.textContent).toBe('Count(12): 2');

    act(() => screen.getByTestId("refetch-12").click())

    await waitFor(async () => {
      expect(count12Dob.textContent).toBe('<updated-dob> [1]');
      expect(countListElement.textContent).toBe('Count(list): 2');
      expect(count10Element.textContent).toBe('Count(10): 1');
      expect(count11Element.textContent).toBe('Count(11): 1');
      expect(count12Element.textContent).toBe('Count(12): 2');

      expect(countProfileUsername.textContent).toBe('Count(12): 2');
      expect(countProfileDob.textContent).toBe('Count(12): 2');

    })

    act(() => screen.getByTestId("refetch-12").click())
    await waitFor(async () => {
      expect(count12Dob.textContent).toBe('<updated-dob> [2]');
      expect(countListElement.textContent).toBe('Count(list): 3');
      expect(count10Element.textContent).toBe('Count(10): 1');
      expect(count11Element.textContent).toBe('Count(11): 1');
      expect(count12Element.textContent).toBe('Count(12): 4');

      expect(countProfileUsername.textContent).toBe('Count(12): 2');
      expect(countProfileDob.textContent).toBe('Count(12): 3');
    })

    act(() => screen.getByTestId("refetch-12").click())
    await waitFor(async () => {
      expect(count12Dob.textContent).toBe('<updated-dob> [4]');
      expect(countListElement.textContent).toBe('Count(list): 4');
      expect(count10Element.textContent).toBe('Count(10): 1');
      expect(count11Element.textContent).toBe('Count(11): 1');
      expect(count12Element.textContent).toBe('Count(12): 6');

      expect(countProfileUsername.textContent).toBe('Count(12): 2');
      expect(countProfileDob.textContent).toBe('Count(12): 4');
    })

  });

  it("Using 'fetch': Should rerender only the component where data changes", async () => {

    function UserListComponent() {

      const fetch = (rand: any) => fakeFetch(Object.values(users), 100)

      const store = useQuery({
        table: "user",
        get: { fetch },
        fields: {
          user: ["id", "dob", "username"],
        }
      })

      const count = useRenderCount([store])

      const renderItem = (user: User, index: number) => user && <UserComponent key={index} id={user.id} />

      return (
        <>
          <p data-testid={`isFetching`} >{JSON.stringify(store.isFetching)}</p>
          <p data-testid={`count-list`} >Count(list): {count}</p>
          {store.result?.map(renderItem)}
        </>
      )
    }

    render(
      <>
        <UserListComponent />
        <Profile id={12} />
      </>,
      { wrapper }
    )

    const isFetching = screen.getByTestId("isFetching");
    expect(isFetching.textContent).toBe('true');

    await waitFor(async () => {
      expect(isFetching.textContent).toBe('false');
    })

    const countListElement = screen.getByTestId("count-list");
    const count10Element = screen.getByTestId("count-10");
    const count11Element = screen.getByTestId("count-11");
    const count12Element = screen.getByTestId("count-12");
    const count12Dob = screen.getByTestId("dob-12");

    const countProfileUsername = screen.getByTestId("count-profile-12");
    const countProfileDob = screen.getByTestId("count-dob-12");

    expect(countListElement.textContent).toBe('Count(list): 2');
    expect(count10Element.textContent).toBe('Count(10): 0');
    expect(count11Element.textContent).toBe('Count(11): 0');
    expect(count12Element.textContent).toBe('Count(12): 0');

    expect(countProfileUsername.textContent).toBe('Count(12): 2');
    expect(countProfileDob.textContent).toBe('Count(12): 2');

    act(() => screen.getByTestId("refetch-12").click())

    await waitFor(async () => {
      expect(count12Dob.textContent).toBe('<updated-dob> [0]');
      expect(countListElement.textContent).toBe('Count(list): 3');
      expect(count10Element.textContent).toBe('Count(10): 1');
      expect(count11Element.textContent).toBe('Count(11): 1');
      expect(count12Element.textContent).toBe('Count(12): 3');

      expect(countProfileUsername.textContent).toBe('Count(12): 2');
      expect(countProfileDob.textContent).toBe('Count(12): 2');
    })

    act(() => screen.getByTestId("refetch-12").click())
    await waitFor(async () => {
      expect(count12Dob.textContent).toBe('<updated-dob> [3]');
      expect(countListElement.textContent).toBe('Count(list): 4');
      expect(count10Element.textContent).toBe('Count(10): 1');
      expect(count11Element.textContent).toBe('Count(11): 1');
      expect(count12Element.textContent).toBe('Count(12): 5');

      expect(countProfileUsername.textContent).toBe('Count(12): 2');
      expect(countProfileDob.textContent).toBe('Count(12): 3');
    })

    act(() => screen.getByTestId("refetch-12").click())
    await waitFor(async () => {
      expect(count12Dob.textContent).toBe('<updated-dob> [5]');
      expect(countListElement.textContent).toBe('Count(list): 5');
      expect(count10Element.textContent).toBe('Count(10): 1');
      expect(count11Element.textContent).toBe('Count(11): 1');
      expect(count12Element.textContent).toBe('Count(12): 7');

      expect(countProfileUsername.textContent).toBe('Count(12): 2');
      expect(countProfileDob.textContent).toBe('Count(12): 4');
    })

  });

  it("The parent component should not rerender", async () => {

    function UpdateDob(props: { id: number; }) {

      const data: Partial<User> = {
        id: props.id,
        dob: "21-02-1998",
      }

      const fetch = (rand: any) => fakeFetch({ ...data, dob: `<updated-dob> [${rand}]` }, 100)

      const store = useQuery({
        table: "user",
        get: {
          result: data,
          fetch,
          where: { id: props.id }
        },
        getData: res => res,
        fields: { user: ["dob"] },
        enabled: false
      })

      const count = useRenderCount()

      const handleButtonClick = () => store.refetch(count);

      return (
        <>
          <p data-testid={`count-dob`} >Count dob: {count}</p>
          <p data-testid="dob">{store.result?.dob}</p>
          <button data-testid={`refetch`} onClick={handleButtonClick}>Update</button>
        </>
      )
    }

    function Parent() {

      const count = useRenderCount()


      return (
        <>
          <p data-testid={`count-parent`} >Count parent: {count}</p>
          <UpdateDob id={12} />
        </>
      )
    }

    render(<Parent />, { wrapper })

    const countParent = screen.getByTestId("count-parent");
    const countDob = screen.getByTestId("count-dob");
    const refetchDob = screen.getByTestId("refetch");


    expect(countParent.textContent).toBe('Count parent: 0');
    expect(countDob.textContent).toBe('Count dob: 1');


    act(() => refetchDob.click())

    await waitFor(async () => {
      expect(countParent.textContent).toBe('Count parent: 0');
      expect(countDob.textContent).toBe('Count dob: 3');
    })

    act(() => refetchDob.click())

    await waitFor(async () => {
      expect(countParent.textContent).toBe('Count parent: 0');
      expect(countDob.textContent).toBe('Count dob: 5');
    })

  });
});


describe("useMutation hook", () => {


  it("useMutation should mutate a value.", async () => {

    const data: Partial<User> = {
      id: 12,
      dob: "21-02-1998",
    }

    const fetch = (next: { dob: string; }) => fakeFetch({ ...data, ...next }, 100)

    const { result } = renderHook(() => (
      useMutation({
        table: "user",
        mutate: fetch,
      })
    ), { wrapper });

    const response = await act(() => result.current.mutate({ dob: "01-01-2000" }));

    expect(response).toStrictEqual({ id: 12, dob: "01-01-2000" });

    await waitFor(() => expect(result.current.isLoading).toBe(false))
  });

})


describe("useInfiniteQuery hook", () => {


  it("Pagination test 1", async () => {

    function UserListComponent() {

      const fetch = async (nextParams: any) => fakePaginatingFetch(Object.values(users), nextParams)

      const store = useInfiniteQuery({
        table: "user",
        get: { fetch: next => fetch(next) },
        getData: (result) => result.data,
        getNextPageParams: (result) => result.nextParams,
        getNextPageKey: (result) => !result.nextParams ? null : result.nextParams.createdAt.toString(),
        fields: {
          user: ["id", "dob", "username"],
        }
      })

      const getNextPage = () => store.fetchNextPage()
      const refetch = () => store.refetch()

      return (
        <>
          <p data-testid={`result`}>{JSON.stringify(store.result?.length)}</p>
          <button data-testid={`refetch`} onClick={refetch}></button>
          <button data-testid={`fetchNextPage`} onClick={getNextPage}></button>
        </>
      )
    }

    render(<UserListComponent />, { wrapper })

    const refetch = screen.getByTestId("refetch");
    const fetchNextPage = screen.getByTestId("fetchNextPage");
    const result = screen.getByTestId("result");

    await waitFor(async () => {
      expect(result.textContent).toBe('2');
    })

    act(() => fetchNextPage.click())

    await waitFor(async () => {
      expect(result.textContent).toBe('3');
    })

    act(() => refetch.click())

    await waitFor(async () => {
      expect(result.textContent).toBe('3');
    })
  });

  it("Pagination test 2", async () => {

    function UserListComponent() {

      const fetch = async (nextParams: any) => {
        const result = await fakePaginatingFetch(Object.values(users), nextParams)
        return result
      }

      const store = useInfiniteQuery({
        table: "user",
        get: { fetch: next => fetch(next) },
        getData: (result) => result.data,
        getNextPageParams: (result) => result.nextParams,
        getNextPageKey: (result) => !result.nextParams ? null : result.nextParams.createdAt.toString(),
        fields: {
          user: ["id", "dob", "username"],
        }
      })

      const getNextPage = () => store.fetchNextPage()
      const refetch = () => store.refetch()

      return (
        <>
          <p data-testid={`result`}>{JSON.stringify(store.result?.length)}</p>
          <button data-testid={`refetch`} onClick={refetch}></button>
          <button data-testid={`fetchNextPage`} onClick={getNextPage}></button>
        </>
      )
    }

    render(<UserListComponent />, { wrapper })

    const refetch = screen.getByTestId("refetch");
    const result = screen.getByTestId("result");

    await waitFor(async () => {
      expect(result.textContent).toBe('2');
    })

    act(() => refetch.click())

    await waitFor(async () => {
      expect(result.textContent).toBe('2');
    })
  });

})