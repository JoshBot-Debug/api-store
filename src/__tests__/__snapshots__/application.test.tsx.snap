// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useInfiniteQuery hook Join tests: Join on hasMany field 1`] = `
Object {
  "id": 10,
  "images": Array [
    Object {
      "id": 54,
      "thumbnails": Array [
        Object {
          "height": 128,
          "id": 206,
        },
        Object {
          "height": 256,
          "id": 207,
        },
        Object {
          "height": 512,
          "id": 208,
        },
        Object {
          "height": 720,
          "id": 209,
        },
        Object {
          "height": 1728,
          "id": 210,
        },
      ],
    },
    Object {
      "id": 55,
      "thumbnails": Array [
        Object {
          "height": 128,
          "id": 211,
        },
        Object {
          "height": 256,
          "id": 212,
        },
        Object {
          "height": 512,
          "id": 213,
        },
        Object {
          "height": 720,
          "id": 214,
        },
        Object {
          "height": 1656,
          "id": 215,
        },
      ],
    },
    Object {
      "id": 56,
      "thumbnails": Array [
        Object {
          "height": 128,
          "id": 216,
        },
        Object {
          "height": 256,
          "id": 217,
        },
        Object {
          "height": 512,
          "id": 218,
        },
        Object {
          "height": 720,
          "id": 219,
        },
        Object {
          "height": 1728,
          "id": 220,
        },
      ],
    },
    Object {
      "id": 57,
      "thumbnails": Array [
        Object {
          "height": 128,
          "id": 221,
        },
        Object {
          "height": 256,
          "id": 222,
        },
        Object {
          "height": 512,
          "id": 223,
        },
        Object {
          "height": 720,
          "id": 224,
        },
        Object {
          "height": 1728,
          "id": 225,
        },
      ],
    },
  ],
  "user": 2,
}
`;

exports[`useInfiniteQuery hook Join tests: Join on hasMany field alt way 1`] = `
Object {
  "id": 10,
  "images": Array [
    Object {
      "id": 54,
      "thumbnails": Array [
        206,
        207,
        208,
        209,
        210,
      ],
    },
    Object {
      "id": 55,
      "thumbnails": Array [
        211,
        212,
        213,
        214,
        215,
      ],
    },
    Object {
      "id": 56,
      "thumbnails": Array [
        216,
        217,
        218,
        219,
        220,
      ],
    },
    Object {
      "id": 57,
      "thumbnails": Array [
        221,
        222,
        223,
        224,
        225,
      ],
    },
  ],
  "user": 2,
}
`;

exports[`useInfiniteQuery hook Join tests: Join on hasMany field by force 1`] = `
Object {
  "id": 10,
  "images": Object {
    "id": 43,
    "thumbnails": Array [
      161,
      162,
      163,
      164,
      165,
    ],
  },
  "user": 2,
}
`;

exports[`useInfiniteQuery hook Join tests: Join on hasMany field with another hasMany field 1`] = `
Object {
  "id": 10,
  "images": Array [
    Object {
      "id": 54,
      "thumbnails": Array [
        Object {
          "height": 128,
          "id": 206,
        },
        Object {
          "height": 256,
          "id": 207,
        },
        Object {
          "height": 512,
          "id": 208,
        },
        Object {
          "height": 720,
          "id": 209,
        },
        Object {
          "height": 1728,
          "id": 210,
        },
      ],
    },
    Object {
      "id": 55,
      "thumbnails": Array [
        Object {
          "height": 128,
          "id": 211,
        },
        Object {
          "height": 256,
          "id": 212,
        },
        Object {
          "height": 512,
          "id": 213,
        },
        Object {
          "height": 720,
          "id": 214,
        },
        Object {
          "height": 1656,
          "id": 215,
        },
      ],
    },
    Object {
      "id": 56,
      "thumbnails": Array [
        Object {
          "height": 128,
          "id": 216,
        },
        Object {
          "height": 256,
          "id": 217,
        },
        Object {
          "height": 512,
          "id": 218,
        },
        Object {
          "height": 720,
          "id": 219,
        },
        Object {
          "height": 1728,
          "id": 220,
        },
      ],
    },
    Object {
      "id": 57,
      "thumbnails": Array [
        Object {
          "height": 128,
          "id": 221,
        },
        Object {
          "height": 256,
          "id": 222,
        },
        Object {
          "height": 512,
          "id": 223,
        },
        Object {
          "height": 720,
          "id": 224,
        },
        Object {
          "height": 1728,
          "id": 225,
        },
      ],
    },
  ],
  "user": 2,
}
`;

exports[`useInfiniteQuery hook Join tests: Join on hasMany field, multiple by force 1`] = `
Object {
  "id": 10,
  "images": Array [
    Object {
      "id": 43,
      "thumbnails": Array [
        161,
        162,
        163,
        164,
        165,
      ],
    },
    Object {
      "id": 44,
      "thumbnails": Array [
        166,
        167,
        168,
        169,
        170,
      ],
    },
  ],
  "user": 2,
}
`;
